decision-records/006-javascript-compatibility.md

```diff
@@ -0,0 +1,161 @@
+
+The inline script adding the `govuk-frontend-supported` class cannot be run in a `<script type="module">`. This would defer the execution of the script adding the class until after the page loads. This defeats the purpose of that script, as it needs to run immediately so it prevents flashes of non-interactive styles.
+
+### Publishing our source code
+
+We could publish our ES2015 source code, without making any transformations.
+
+While we could lint away some compatibility concerns, this would require more manual testing to ensure we match what the browsers we support can parse and execute.
+
+Due to the more manual nature, it would also be more prone to bugs and regressions.
+
+And finally, due to the extra knowledge necessary to know what's compatible or not, it would require extra effort from developers and contributors to offer changes.
+
+### Guarding based on newer APIs and letting JavaScript error
+
+Instead of using the availability of `HTMLScriptElement.prototype.noModule` to set the `govuk-frontend-supported` class (which governs both styling and execution of our components), we could:
+
+- pick a more modern feature to check (for example `Promise.prototype.finally`)
+- transpile down to the more modern set of browsers that support it (rather than all the way down to ES2015)
+- and let JavaScript error in browsers that support `type="module"` but not the newer features that appear in the code we ship
+
+This would further reduce the amount of features that get transpiled (most notable would be spread `...` and `async/await` if we chose `Promise.prototype.finally`). However this comes with the following drawbacks:
+
+- The extra errors in browsers that support `type="module"` but not our new baseline makes it harder to know if the error is expected (because the browser is not supported) or not. This complicates both the debugging for our users and the triaging of bugs for us.
+- Because of syntax errors, service code bundled with GOV.UK Frontend would not get a chance to be executed at all. This would require extra effort on the services part to decide what can be bundled with GOV.UK Frontend and what needs to be in a separate script to ensure it runs.
+
+### Recommending dynamic import of GOV.UK Frontend
+
+Using a dynamic `import()` call would allow to handle syntax errors when parsing GOV.UK Frontend (via a `.catch` on the promise, or a `try/catch` block, depending on the browsers you'd want to support).
+
+This complicates the integration for services as it requires more JavaScript knowledge to understand which piece of code will run or not if GOV.UK Frontend fails to parse. It also complicates the import of multiple individual components, requiring to make sure the `import()` calls run in parallel to not hinder performance (on top of the individual components not being discoverable by [the browsers' preload-scanner](https://web.dev/preload-scanner/)), but handle the potential failure of these calls appropriately.
```

Comment:
Do we need to mention the opposite?

That as our JavaScript takes longer and longer to run, we'll delay the initialisation of components that run last (as they get queried, looped and have `init()` run)

Reply:
What do you mean by "the opposite"? I'm sure I understand what leads to "our JavaScript takes longer and longer to run" (only thing I come to mind is the growing number of transformations that will make our file size grow).

Reply:
Sorry I was commenting on "`import()` calls run in parallel"

As in, do we need to comment on "the opposite" benefits of Promises?

1. By supporting sequential `.init()`, we can simplify integration by avoiding Promises
2. By supporting concurrent `.init()`, we can avoid render blocking from slow `forEach()`

Or shall we remove anything that can be moved to an API decision record?